server/src/main/java/org/opensearch/search/aggregations/bucket/FastFilterRewriteHelper.java

```diff
@@ -439,61 +404,319 @@ public static boolean tryFastFilterAggregation(
-        for (i = 0; i < filters.length; i++) {
-            if (counts[i] > 0) {
-                long bucketKey = i; // the index of filters is the key for filters aggregation
-                if (fastFilterContext.aggregationType instanceof AbstractDateHistogramAggregationType) {
-                    final DateFieldMapper.DateFieldType fieldType =
-                        ((AbstractDateHistogramAggregationType) fastFilterContext.aggregationType).getFieldType();
-                    bucketKey = fieldType.convertNanosToMillis(
-                        NumericUtils.sortableBytesToLong(((PointRangeQuery) filters[i].getQuery()).getLowerPoint(), 0)
-                    );
-                }
-                incrementDocCount.accept(bucketKey, counts[i]);
-                s++;
-                if (s > size) {
+        long[][] ranges = new long[bucketCount][2];
+        if (bucketCount > 0) {
+            roundedLow = preparedRounding.round(fieldType.convertNanosToMillis(low));
+
+            int i = 0;
+            while (i < bucketCount) {
+                // Calculate the lower bucket bound
+                long lower = i == 0 ? low : fieldType.convertRoundedMillisToNanos(roundedLow);
+                roundedLow = preparedRounding.round(roundedLow + interval);
+
+                // Subtract -1 if the minimum is roundedLow as roundedLow itself
+                // is included in the next bucket
+                long upper = i + 1 == bucketCount ? high : fieldType.convertRoundedMillisToNanos(roundedLow) - 1;
+
+                ranges[i][0] = lower;
+                ranges[i][1] = upper;
+                i++;
+            }
+        }
+
+        return ranges;
+    }
+
+    private static DebugInfo multiRangesTraverse(
+        final PointValues.PointTree tree,
+        final long[][] ranges,
+        final BiConsumer<Long, Integer> incrementDocCount,
+        final DateFieldMapper.DateFieldType fieldType,
+        final int size
+    ) throws IOException {
+        // ranges are connected and in ascending order
+        Iterator<long[]> rangeIter = Arrays.stream(ranges).iterator();
+        long[] activeRange = rangeIter.next();
+
+        // make sure the first range at least crosses the min value of the tree
+        boolean noRangeMatches = false;
+        if (activeRange[0] > NumericUtils.sortableBytesToLong(tree.getMaxPackedValue(), 0)) {
+            noRangeMatches = true;
+        } else {
+            while (activeRange[1] < NumericUtils.sortableBytesToLong(tree.getMinPackedValue(), 0)) {
+                if (rangeIter.hasNext()) {
+                    activeRange = rangeIter.next();
+                } else {
+                    noRangeMatches = true;
                     break;
                 }
             }
         }
+        DebugInfo debugInfo = new DebugInfo();
+        if (noRangeMatches) {
+            logger.debug("No ranges match the query, skip the fast filter optimization");
+            return debugInfo;
+        }
```

Comment:
Can simplify this a bit:

```
        DebugInfo debugInfo = new DebugInfo();
        if (activeRange[0] > NumericUtils.sortableBytesToLong(tree.getMaxPackedValue(), 0)) {
            logger.debug("No ranges match the query, skip the fast filter optimization");
            return debugInfo;
        }
        while (activeRange[1] < NumericUtils.sortableBytesToLong(tree.getMinPackedValue(), 0)) {
                if (!rangeIter.hasNext()) {
                    logger.debug("No ranges match the query, skip the fast filter optimization");
                    return debugInfo;
                }
                activeRange = rangeIter.next();
        }
```

Reply:
Or even simpler would be (relies on the fact that activeRange[0] increases monotonically, so if activeRange[0] were greater than MaxPackedValue, the final will also be greater):

```
DebugInfo debugInfo = new DebugInfo();
while (activeRange[1] < NumericUtils.sortableBytesToLong(tree.getMinPackedValue(), 0) && rangeIter.hasNext()) {
                activeRange = rangeIter.next();
}
if (activeRange[0] > NumericUtils.sortableBytesToLong(tree.getMaxPackedValue(), 0) || activeRange[1] < NumericUtils.sortableBytesToLong(tree.getMinPackedValue(), 0)) {
            logger.debug("No ranges match the query, skip the fast filter optimization");
            return debugInfo;
}
```

Reply:
Thanks, I like the first simplification, updated.

Reply:
Just wondering if there is reason for not picking the second. Does look bit more compact compared to first!?

Reply:
My thinking is to make the code also straightforward to read, so also easy to change if needed one day. Imagine later on someone needs to spend several minutes to realise it actually works correctly. Or the first one, people just need read through and understand directly.